composeApp/src/desktopMain/kotlin/org/ooni/probe/update/DesktopUpdateController.kt

```diff
@@ -0,0 +1,128 @@
+                            Logger.w("Appcast URL: ${UpdateConfig.URL}")
+                        }
+                        -999 -> {
+                            Logger.e("Update system setup or validation error: ${error.message}")
+                        }
+                        else -> {
+                            Logger.w("General update error (${error.code}): ${error.message}")
+                        }
+                    }
+                }
+
+                updateManager.setStateCallback { s ->
+                    Logger.d("Update system state: $s")
+                    _state.value = s
+                    _error.value = updateManager.getLastError()
+                }
+
+                Logger.i("Initializing update system...")
+                Logger.i("Appcast URL: ${UpdateConfig.URL}")
+                Logger.i("Public key configured: ${UpdateConfig.PUBLIC_KEY.isNotEmpty()}")
+
+                updateManager.initialize(UpdateConfig.URL, UpdateConfig.PUBLIC_KEY)
+                updateManager.setAutomaticUpdatesEnabled(true)
+                updateManager.setUpdateCheckInterval(24)
+            } catch (e: Exception) {
+                Logger.e("Failed to initialize update system: $e")
+                _error.value = UpdateError(-999, "Setup failed: ${e.message}", "setup")
+                _state.value = UpdateState.ERROR
+            }
+        }
+    }
+
+    fun registerShutdownHandler(appScope: ApplicationScope) {
+        if (updateManager is WinSparkleUpdateManager) {
+            updateManager.setShutdownCallback {
+                Logger.i("WinSparkle requested application shutdown for update installation")
+                appScope.exitApplication()
+            }
+        }
+    }
+
+    fun checkNow() {
+        when {
+            _error.value != null -> {
+                Logger.i("Retrying update check after error")
+                updateManager.retryLastOperation()
+            }
+            updateManager.isHealthy() -> {
+                updateManager.checkForUpdates(showUI = true)
+            }
+            else -> {
+                Logger.w("Update system unhealthy, checking for updates anyway")
+                updateManager.checkForUpdates(showUI = true)
+            }
+        }
+    }
+
+    fun retryLastOperation() {
+        updateManager.retryLastOperation()
+    }
+
+    fun getMenuText(): String =
+        when {
+            _error.value != null -> "Retry Update Check (Error: ${_error.value?.code})"
+            _state.value == UpdateState.CHECKING_FOR_UPDATES -> "Checking for Updates..."
+            _state.value == UpdateState.UPDATE_AVAILABLE -> "Update Available!"
+            _state.value == UpdateState.NO_UPDATE_AVAILABLE -> "Check for Updates (Up to date)"
+            _state.value == UpdateState.INITIALIZING -> "Initializing Updates..."
+            _state.value == UpdateState.ERROR -> "Update System Error - Retry"
+            else -> "Check for Updates..."
+        }
```

Comment:
We need to localize all these messages, either here, or in the UI code. Could be in a separate ticket.